src/JwtCredentialsManager.ts

```diff
@@ -65,36 +67,35 @@ export default abstract class JwtCredentialsManager<
       throw new Error('JwtCredentialManager: Unauthenticated');
     }
 
-    const credentials = await this.getCredentialsFromAuthorization(token);
+    this.credentials = Promise.resolve(
+      this.getCredentialsFromAuthorization(token),
+    );
+    await this.credentials;
 
     // Avoid race conditions with multiple updates.
     if (this.token !== token) {
       return;
     }
 
-    this.credentials = credentials;
-
     // TODO: Don't schedule renewal if the new credentials are expired or
     // almost expired.
     this.scheduleRenewCredentials();
   }
 
-  scheduleRenewCredentials() {
+  async scheduleRenewCredentials() {
     if (this.renewHandle) {
       clearTimeout(this.renewHandle);
     }
 
     const { tokenExpirationMarginSeconds } = this.config;
-    if (tokenExpirationMarginSeconds === null) {
+    if (tokenExpirationMarginSeconds === null || !this.credentials) {
       return;
     }
 
-    const { credentials } = this;
-    if (!credentials) {
-      return;
-    }
+    const resolvedCredentials = await this.credentials;
+    if (!resolvedCredentials) return;
```

Comment:
**nitpick:** to make naming a bit more clear:

```suggestion
    const credentials = await this.credentialsPromise;
    if (!credentials) return;
```

also may want to make formatting match the above early return block